pkg/initializer/initializer.go

```diff
@@ -0,0 +1,180 @@
+		crdClient:       crdClient,
+		namespaceClient: namespaceClient,
+		migrationClient: migrationGetter.StorageVersionMigrations(namespaceName),
+	}
+}
+
+const (
+	// TODO: get the namespace name from the enviroment variable
+	namespaceName   = "kube-storage-migration"
+	singularCRDName = "storageversionmigration"
+	pluralCRDName   = "storageversionmigrations"
+	kind            = "StorageVersionMigration"
+	listKind        = "StorageVersionMigrationList"
+)
+
+func migrationCRD() *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s.%s", pluralCRDName, migrationv1alpha1.GroupName),
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   migrationv1alpha1.GroupName,
+			Version: migrationv1alpha1.SchemeGroupVersion.Version,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural:   pluralCRDName,
+				Singular: singularCRDName,
+				Kind:     kind,
+				ListKind: listKind,
+			},
+			Scope: apiextensionsv1beta1.NamespaceScoped,
+			Subresources: &apiextensionsv1beta1.CustomResourceSubresources{
+				Status: &apiextensionsv1beta1.CustomResourceSubresourceStatus{},
+			},
+		},
+	}
+}
+
+func migrationForResource(resource schema.GroupVersionResource) *migrationv1alpha1.StorageVersionMigration {
+	var name string
+	if len(resource.Group) != 0 {
+		name = fmt.Sprintf("%s.%s.%s-", resource.Group, resource.Version, resource.Resource)
+	} else {
+		name = fmt.Sprintf("%s.%s-", resource.Version, resource.Resource)
+	}
+	return &migrationv1alpha1.StorageVersionMigration{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: name,
+		},
+		Spec: migrationv1alpha1.StorageVersionMigrationSpec{
+			Resource: migrationv1alpha1.GroupVersionResource{
+				Group:    resource.Group,
+				Version:  resource.Version,
+				Resource: resource.Resource,
+			},
+		},
+	}
+}
+
+func (init *initializer) initializeCRD() error {
+	crdName := fmt.Sprintf("%s.%s", pluralCRDName, migrationv1alpha1.GroupName)
+	// check if crd already exists
+	_, err := init.crdClient.Get(crdName, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+	if err != nil && errors.IsNotFound(err) {
+		_, err := init.crdClient.Create(migrationCRD())
+		return err
+	}
+
+	// delete the crd, and wait for it's deletion
+	if err := init.crdClient.Delete(crdName, nil); err != nil {
+		return err
+	}
+	err = wait.PollImmediate(500*time.Millisecond, 30*time.Second, func() (bool, error) {
+		_, err := init.crdClient.Get(crdName, metav1.GetOptions{})
+		if err == nil {
+			return false, nil
+		}
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	})
+	if err != nil {
+		return err
+	}
+	_, err = init.crdClient.Create(migrationCRD())
+	return err
+}
+
+// TODO: remove this function. Users will use provided yaml files to create the
+// namespace, and then create the initializer in the namespace.
```

Comment:
it seems that the whole CRD creation etc could be done out of band, like what you suggested above for namespace